src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -0,0 +1,259 @@
+            state.RequestMessage = request;
+            state.Handler = this;
+
+            Task.Factory.StartNew(
+                s =>
+                {
+                    var rcs = (RequestCompletionSource)s;
+                    rcs.Handler.StartRequest(rcs);
+                },
+                state,
+                CancellationToken.None,
+                TaskCreationOptions.DenyChildAttach,
+                TaskScheduler.Default);
+
+            return state.Task;
+        }
+
+        #region Private methods
+
+        private void StartRequest(RequestCompletionSource state)
+        {
+            HttpResponseMessage responseMessage = null;
+            SafeCurlHandle requestHandle = null;
+
+            if (state.CancellationToken.IsCancellationRequested)
+            {
+                state.TrySetCanceled();
+                return;
+            }
+
+            var cancellationTokenRegistration = state.CancellationToken.Register(x => ((RequestCompletionSource) x).TrySetCanceled(), state);
+
+            try
+            {
+                requestHandle = CreateRequestHandle(state);
+
+                state.CancellationToken.ThrowIfCancellationRequested();
+                int result = Interop.libcurl.curl_easy_perform(requestHandle);
+                if (CURLcode.CURLE_OK != result)
+                {
+                    throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result));
+                }
+
+                // TODO: Handle requests with body
+
+                state.CancellationToken.ThrowIfCancellationRequested();
+                responseMessage = CreateResponseMessage(requestHandle, state.RequestMessage);
+                state.TrySetResult(responseMessage);
+            }
+            catch (Exception ex)
+            {
+                HandleAsyncException(state, ex);
+            }
+            finally
+            {
+                SafeCurlHandle.DisposeAndClearHandle(ref requestHandle);
+                cancellationTokenRegistration.Dispose();
+            }
+        }
+
+        private void SetOperationStarted()
+        {
+            if (!_anyOperationStarted)
+            {
+                _anyOperationStarted = true;
+            }
+        }
+
+        private SafeCurlHandle CreateRequestHandle(RequestCompletionSource state)
+        {
+            // TODO: If this impacts perf, optimize using a handle pool
+            SafeCurlHandle requestHandle = Interop.libcurl.curl_easy_init();
+            if (requestHandle.IsInvalid)
+            {
+                throw new HttpRequestException(SR.net_http_client_execution_error);
+            }
+
+            Interop.libcurl.curl_easy_setopt(requestHandle, CURLoption.CURLOPT_URL,
+                state.RequestMessage.RequestUri.OriginalString);
+            if (_automaticRedirection)
+            {
+                Interop.libcurl.curl_easy_setopt(requestHandle, CURLoption.CURLOPT_FOLLOWLOCATION, 1);
```

Comment:
@vijaykota Sorry, there are multiple TODOs but I can't find a specific one regarding configuring automatic redirection in a secure way. (Pushing for this as it's security-related.)